rskj-core/src/main/java/org/ethereum/core/BlockHeader.java

```diff
@@ -471,7 +498,30 @@ public String getParentShortHash() {
         return HashUtil.shortHash(getParentHash().getBytes());
     }
 
-    private static BigInteger parseBigInteger(byte[] bytes) {
-        return bytes == null ? BigInteger.ZERO : BigIntegers.fromUnsignedByteArray(bytes);
+    public byte[] getMiningForkDetectionData() {
+        if(includeForkDetectionData) {
+            if (hasMiningFields()) {
+                byte[] encodedBlock = getEncoded(false, false);
+                byte[] hashForMergedMining = HashUtil.keccak256(encodedBlock);
+
+                byte[] hashForMergedMiningPrefix = Arrays.copyOfRange(hashForMergedMining, 0, HASH_FOR_MERGED_MINING_PREFIX_LENGTH);
+                byte[] coinbaseTransaction = getBitcoinMergedMiningCoinbaseTransaction();
+
+                List<Byte> hashForMergedMiningPrefixAsList = Arrays.asList(ArrayUtils.toObject(hashForMergedMiningPrefix));
+                List<Byte> coinbaseAsList = Arrays.asList(ArrayUtils.toObject(coinbaseTransaction));
+
+                int position = Collections.lastIndexOfSubList(coinbaseAsList, hashForMergedMiningPrefixAsList);
+                if (position == -1) {
```

Comment:
If we're parsing the mining fields, then maybe it would be wise to cache also the RSKTAG and the ProofOfWork rule wouldn't need to compute this again (it would only call header.getRSKTAG() or something). 
I don't like much that the tag-handling code be duplicated in two classes, being this critical to the consensus and security of the platform.

Reply:
In this case, the value extracted is not the same as the one in PowRule. 
I think that BlockHeader may be responsible for knowing how to extract the values used in PowRule (this a topic I consider open for discussion). 
On this feature implementation, a big one, I'd rather not change PowRule and leave that for a future version. We all know that PowRule needs a huge refactor.